Docker/run.sh

```diff
@@ -0,0 +1,64 @@
+#!/bin/bash
+# This script prepare the host environment for running one CRAB service based on CRAB
+# TaskWorker container image and runs it
+# Currently those services include TaskWorker and Publisher_*
+
+# This script is meant to be called in the Dockerfile CMD, and whoever issues the docker run command
+# must pass the SERVICE name as an environmental variable  via the '-e SERVICE=name' option
+# The script expects to be run in the /data/srv/$SERVICE directory and to in find there the start.sh
+# script appropriate for the $SERVICE being run
+
+# cleanup environment
+unset X509_USER_PROXY
+unset X509_USER_CERT
+unset X509_USER_KEY
+
+# ensure container has needed mounts
+check_link(){
+# function checks if symbolic links required to start service exists and if they are not broken
+
+  if [ -L $1 ] ; then
+    if [ -e $1 ] ; then
+       return 0
+    else
+       unlink $1
+       return 1
+    fi
+  else
+    return 1
+  fi
+}
+
+# directories/files that should be created before starting the container :
+# note a trick to do partial string match from https://unix.stackexchange.com/a/465906
+if [[ $SERVICE == TaskWorker ]]; then
+  # -/data/hostdisk/${SERVICE}/cfg/TaskWorkerConfig.py
+  # -/data/hostdisk/${SERVICE}/logs
+  declare -A links=( ["current/TaskWorkerConfig.py"]="/data/hostdisk/${SERVICE}/cfg/TaskWorkerConfig.py" ["logs"]="/data/hostdisk/${SERVICE}/logs" ["nohup.out"]="/data/hostdisk/${SERVICE}/nohup.out")
+elif [[ $SERVICE == Publisher* ]]; then
+  # -/data/hostdisk/${SERVICE}/cfg/PublisherConfig.py
+  # -/data/hostdisk/${SERVICE}/logs
+  # -/data/hostdisk/${SERVICE}/PublisherFiles
+  declare -A links=( ["PublisherConfig.py"]="/data/hostdisk/${SERVICE}/cfg/PublisherConfig.py" ["logs"]="/data/hostdisk/${SERVICE}/logs" ["/data/srv/Publisher_files"]="/data/hostdisk/${SERVICE}/PublisherFiles" ["nohup.out"]="/data/hostdisk/${SERVICE}/nohup.out")
+fi
+
+for name in "${!links[@]}";
+do
+  check_link "${name}" || ln -s "${links[$name]}" "$name"
+done
+
+# if GH repositories location is not already defined, set a default
+if ! [ -v GHrepoDir ]
+then
+  GHrepoDir='/data/hostdisk/repos'
+fi
+
+# cd to the SERVICE directory
```

Comment:
Why do we need this part?  Working directory is already set here: https://github.com/dmwm/CRABServer/blob/master/src/script/Container/runContainer.sh#L51 at container start time. 

one more thing, i cannot find change which would download `run.sh` file to 'the needed directory'. It should be added here: https://github.com/dmwm/CRABServer/blob/master/Docker/install.sh#L28-L29

Reply:
you are too fast.
I am not sure what  I am doing with working directory. I got convinced that this script will run with /data as cwd. Indeed I had looked at runContainer but did not notice the second option in DOCKER_OPT :-(

As to how to get it there, I thought that it is better to move run.sh to Docker directory and will put it in the image via ADD (IIRC) like done for crabserver. Since it is same script for both Publisher and TW.. and I have a secret idea to use this also to run WMArchiveUploader and run it as a container on the schedd's, instead of maintaining there a WMCore clone which periodically gets obsolete.

Reply:
OTOH, since we pass $SERVICE to the container, the container can figure out where to run. Why should this internal detail be exposed to who issues docker run ?

Reply:
when we enter container, we want it to start at `/data/srv/${DIRECTORY}` directory. That is what `-w` option does [1].  otherwise it will be [`/data`](https://github.com/dmwm/CRABServer/blob/master/Docker/Dockerfile#L10) directory and one will have to `cd ` manually. I don't know how else to achieve result of:

> I run `docker exec -it Publisher_schedd  bash` and end up in `/data/srv/Publisher` OR I run `docker exec -it TaskWorker  bash` and end up in `/data/srv/TaskManager`

Usually working directory is set in Dockerfile, e.g. https://github.com/dmwm/CMSKubernetes/blob/master/docker/crabserver/Dockerfile#L30, but in our case it is impossible: we start 2 services using one Dockerfile, so we cannot define this variable in Dockerfile to be `ENV WDIR=/data/srv/Publisher` or `ENV WDIR=/data/srv/TaskManager`.

[1] scroll until the very end for documentation: https://docs.docker.com/engine/reference/run/

Reply:
Got it thanks. I had forgotten the `docker exec it $SERVICE /bin/bash` . Current code in my branch should do as you say. I will polish a bit more and test.